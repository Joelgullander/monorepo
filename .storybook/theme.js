import { create } from '@storybook/theming';

const theme = create({
    base: 'light',
  
    colorPrimary: '#D52B1E',
    colorSecondary: '#D52B1E',
  
    // UI
    appBg: 'white',
    appContentBg: '#f9f9f9',
    appBorderColor: '#f000',
    appBorderRadius: 0,
  
    // Typography
    fontBase: '"Roboto", sans-serif',
    fontCode: 'monospace',
  
    // Text colors
    textColor: '#212121',
    textInverseColor: 'rgba(255,255,255,0.9)',
  
    // Toolbar default and active colors
    barTextColor: 'white',
    barSelectedColor: '#D52B1E',
    barBg: '#000',
  
    // Form colors
    inputBg: 'white',
    inputBorder: 'silver',
    inputTextColor: 'black',
    inputBorderRadius: 4,
  
    brandTitle: 'Hiab components',
    brandUrl: 'https://hiab.com',
    brandImage: 'https://www.hiab.com/Content/images/logotypes/hiab2x2.png',
  });

  export default theme;