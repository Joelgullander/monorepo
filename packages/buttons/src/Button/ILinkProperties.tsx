export interface ILinkProperties {
    url: string,
    internal: boolean,
};
