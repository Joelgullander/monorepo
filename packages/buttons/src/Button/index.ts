export * from './Button';
export * from './TextLink';
