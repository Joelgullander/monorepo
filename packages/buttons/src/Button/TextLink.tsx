import React from 'react';
import { Link } from 'react-router-dom';
import * as style from './TextLink.style';
import { IIcon } from '@hiab/icons';
import { ILinkProperties } from './ILinkProperties';

// export type Underline = 'solid' | 'dotted' | 'none' | 'half-solid';

export enum Underline {
  SOLID = 'solid',
  DOTTED = 'dotted',
  NONE = 'none',
  HALF_SOLID = 'half-solid',
}

interface ILinkText {
  link: ILinkProperties,
  icon?: IIcon,
  underline: Underline,
};

export const TextLink: React.FC<ILinkText> = ({
  link,
  children,
  icon,
  underline
}) => {
  if (link.internal) {
    return (
      <style.TextLink
        as={Link}
        to='/'
        underline={underline}
      >
        {icon && icon}
        {children}
      </style.TextLink>
    );
  } else {
    return (
      <style.TextLink
        as='a'
        href={link.url}
        underline={underline}
      >
        {icon && icon}
        {children}
      </style.TextLink>
    )
  }
}

TextLink.defaultProps = {
  underline: Underline.NONE,
};
