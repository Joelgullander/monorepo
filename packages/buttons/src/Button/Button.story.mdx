import { Meta, Story, Props, Preview } from '@storybook/addon-docs/blocks';
import { withKnobs, text, boolean, number } from "@storybook/addon-knobs";
import { action } from '@storybook/addon-actions';
import { Icon } from '@hiab/icons';

import { Button, LinkButton } from './Button';
import { TextLink } from './TextLink';

<Meta title='Components|Buttons' component={Button} decorators={[withKnobs]} />

# Button

### Button / link option

<Preview>
    <Story name="button link">
        <TextLink
            link={{
                url: "https://hiab.com",
                internal: false
            }}
            icon={<Icon name="downArrow" color="#D52B1E" width="12px" height="12px" />}
        >
            Service offering
        </TextLink>
    </Story>
</Preview>

### Text Link Half-solid

<Preview>
    <Story name="half-solid link">
        <TextLink
            link={{
                url: "https://hiab.com",
                internal: false
            }}
            underline='half-solid'
        >
            Read more
        </TextLink>
    </Story>
</Preview>

### Text Link

<Preview>
    <Story name="text link">
        <TextLink
            link={{
                url: "https://hiab.com",
                internal: false
            }}
        >
            This is a link in text
        </TextLink>
    </Story>
</Preview>

### Text Link Solid 

<Preview>
    <Story name="text link solid">
        <TextLink
            link={{
                url: "https://hiab.com",
                internal: false
            }}
            underline="solid"
        >
            This is a link in text
        </TextLink>
    </Story>
</Preview>

### Text Link dotted

<Preview>
    <Story name="text link dotted">
        <TextLink
            link={{
                url: "https://hiab.com",
                internal: false
            }}
            underline="dotted"
        >
            This is a link in text
        </TextLink>
    </Story>
</Preview>

### Big

<Preview>
    <Story name="big">
        <Button size="big" >
            Send me my invitation
        </Button>
    </Story>
    <Story name="big disabled">
        <Button size="big" disabled>
            Send me my invitation
        </Button>
    </Story>
</Preview>

<Preview>
     <Story name="with icon">
        <Button
            size="big"
            icon={<Icon name="plus" color="#fff" width="12px" height="12px" />}
        >
            Send me my invitation
        </Button>
    </Story>
</Preview>

### Medium

<Preview>
    <Story name="medium">
        <Button size="medium" >
            Send me my invitation
        </Button>
    </Story>
    <Story name="medium disabled">
        <Button size="medium"  disabled>
            Send me my invitation
        </Button>
    </Story>
</Preview>

### Small

<Preview>
    <Story name="small">
        <Button size="small" >
            Send me my invitation
        </Button>
    </Story>
   <Story name="small disabled">
        <Button size="small" disabled>
            Send me my invitation
        </Button>
    </Story>
</Preview>

### Inverted

<Preview>
    <Story name="inverted">
        <Button inverted size="small" >
            View product
        </Button>
    </Story>
     <Story name="inverted and disabled">
        <Button inverted size="small"  disabled>
            View product
        </Button>
    </Story>
</Preview>

### Rounded

<Preview>
    <Story name="rounded">
        <Button
            variant="rounded"
            size="big"
            icon={<Icon name="comments" color="#fff" width="28px" height="25px" />}
        >
            Contact us
        </Button>
    </Story>
</Preview>

### Icon Rounded

<Preview>
    <Story name="icon rounded">
        <Button
            variant="icon-rounded"
            size="big"
            icon={<Icon name="comments" color="#fff" width="28px" height="25px" />}
        />
    </Story>
</Preview>

### Right side icon

<Preview>
    <Story name="rounded right icon">
        <Button
            size="big"
            iconPosition="right"
            icon={<Icon name="downArrow" color="#fff" width="12px" height="12px" />}
        >
            Our brands
        </Button>
    </Story>
     <Story name="active">
        <Button
            size="big"
            iconPosition="right"
            icon={<Icon name="downArrow" color="#D52B1E" width="12px" height="12px" />}
            active
        >
            Our brands
        </Button>
    </Story>
</Preview>

### Can be used as a link
#### If link.internal is being passed the component will render with react-router-dom's *Link*, if false normal HTML-Anchor will be rendered.
<Preview>
    <Story name="link button">
        <LinkButton 
            size="big"
            link={{
                url: "https://hiab.com",
                internal: false
            }}
        >
            Click me
        </LinkButton>
    </Story>
</Preview>

## Props

<Props of={Button} />
