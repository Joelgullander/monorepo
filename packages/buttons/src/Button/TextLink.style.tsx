import styled from 'styled-components';
import { HiabTheme } from '@hiab/theme';
import { Underline } from './TextLink';

interface IStyledTextLink {
    underline: Underline,
    theme: HiabTheme,
};

export const TextLink = styled.div`
    display: inline-block;
    position: relative;
    text-decoration: none;
    line-height: 14px;
    font-family: ${(props: IStyledTextLink) => props.theme.fonts.primary};
    transition: ${(props: IStyledTextLink) => props.theme.transition.primary} ;
    color: ${(props: IStyledTextLink) => {
        if (props.underline === Underline.DOTTED || props.underline === Underline.HALF_SOLID)
            return props.theme.colors.black;
        return props.theme.colors.thunderbird;
    }};

    ${(props: IStyledTextLink) => {
        if (props.underline === Underline.DOTTED)
            return `
                background-image: linear-gradient(to right, ${props.theme.colors.thunderbird} 20%, rgb(213,43,30, 0) 0%);
                background-position: bottom;
                background-size: 4px 1px;
                background-repeat: repeat-x;
            `
        if (props.underline === Underline.HALF_SOLID) {
            return `
                font-size: 13px;
                text-transform: uppercase;
                letter-spacing: 1.3px;
                font-weight: 600;
                &:after {
                    content: '';
                    width: 45px;
                    height: 2px;
                    background: ${props.theme.colors.thunderbird};
                    position: absolute;
                    bottom: -4.5px;
                    left: 1px;
                }
            `
        }
        return `border-bottom: 1px ${props.underline} ${props.theme.colors.thunderbird}`
    }};

    & > svg {
        transform: rotateZ(270deg);
        margin-right: 5px;
    };

    &:hover, &:focus {
        color: ${(props: IStyledTextLink) => {
        if (props.underline === Underline.DOTTED || props.underline === Underline.HALF_SOLID)
            return props.theme.colors.thunderbird;
        return props.theme.colors.black;
    }};
        border-bottom: ${(props: IStyledTextLink) => {
        if (props.underline === Underline.DOTTED)
            return `background-image: linear-gradient(to right, black 100%, rgba(255,255,255,0) 0%);`
        return `1px ${props.underline} ${props.theme.colors.black}`;
    }};
    };

    &:focus {
        outline: none;
    };
`;
