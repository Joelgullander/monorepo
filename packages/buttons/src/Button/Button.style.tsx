import styled from 'styled-components';
import { HiabTheme } from '@hiab/theme';
import { IconPosition, Variants, Size } from './Button';

const height = {
    big: "56px",
    medium: "48px",
    small: "34px",
}

interface IStyledButton {
    theme: HiabTheme,
    disabled?: boolean,
    variant?: Variants,
    size?: Size,
    inverted?: boolean,
    iconPosition?: IconPosition,
    active?: boolean,
}

export const Button = styled.div`
    display: inline-flex;
    justify-content: center;
    align-items: center;
    box-sizing: border-box;
    font-family: ${(props: IStyledButton) => props.theme.fonts.primary};
    padding: ${(props: IStyledButton) => props.variant === Variants.ICON_ROUNDED ? '0 15px' : '0px 26px'};
    height: ${(props: IStyledButton) => height[props.size || Size.MEDIUM]};
    background-color: ${(props: IStyledButton) => !props.inverted
        ? props.theme.colors.white
        : props.theme.colors.thunderbird};
    text-decoration: none;
    text-transform: uppercase;
    transition: ${(props: IStyledButton) => props.theme.transition.primary};
    font-size: ${(props: IStyledButton) => props.size === Size.SMALL ? '12px' : '13px'};
    letter-spacing: ${(props: IStyledButton) => {
        if (props.size === Size.BIG)
            return '1.5px';
        if (props.size === Size.MEDIUM)
            return '1.3px';
        return '1.2px';
    }};
    background-color: ${(props: IStyledButton) => {
        if (props.disabled && props.inverted)
            return props.theme.colors.white;
        if (props.disabled)
            return props.theme.colors.gainsboro;
        if (props.inverted || props.active)
            return props.theme.colors.white;
        return props.theme.colors.thunderbird;
    }};
    color: ${(props: IStyledButton) => {
        if (props.disabled && props.inverted)
            return props.theme.colors.lightGrey;
        if (props.disabled)
            return props.theme.colors.darkSilver;
        if (props.inverted)
            return props.theme.colors.black;
        if (props.active)
            return props.theme.colors.thunderbird;
        return props.theme.colors.white;
    }};
    border-radius: ${(props: IStyledButton) => {
        if (props.variant === Variants.ROUNDED || props.variant === Variants.ICON_ROUNDED)
            return '30px';
        return '0px';
    }
    };
    pointer-events:  ${(props: IStyledButton) => props.disabled ? 'none' : 'all'};
    border: ${(props: IStyledButton) => {
        if (props.inverted && props.disabled)
            return `1px solid ${props.theme.colors.whiteSmoke}`;
        if (props.inverted)
            return `1px solid ${props.theme.colors.nobel}`
        if (props.inverted)
            return 'none';
        return 'none';
    }};

    & > svg {
        ${(props: IStyledButton) => {
        if (props.variant === Variants.ICON_ROUNDED)
            return 'margin: 0;';

        if (props.iconPosition === IconPosition.LEFT)
            return 'margin-right: 19px;';

        return 'margin-left: 12px';
    }
    }
    };

    &:hover, &:focus {
        ${(props: IStyledButton) => {
        if (!props.active) {
            return (`
                background-color: ${props.inverted ? props.theme.colors.thunderbird : props.theme.colors.mineshaft};
                transition: background-color 0.3s ease;
                color: ${props.inverted && props.theme.colors.white};
                border: ${ props.inverted && `1px solid ${props.theme.colors.thunderbird}`};
            `);
        }
        return;
    }}
    };

    &:focus {
        outline: none;
    };
`;
