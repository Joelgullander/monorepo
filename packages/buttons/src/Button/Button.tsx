import React from 'react';
import { Link } from 'react-router-dom';
import * as style from './Button.style';
import { IIcon } from '@hiab/icons';
import { ILinkProperties } from './ILinkProperties';

export enum IconPosition {
  LEFT = 'left',
  RIGHT = 'right',
};

export enum Variants {
  DEFAULT = "default",
  ROUNDED = "rounded",
  ICON_ROUNDED = "icon-rounded",
}

export enum Size {
  BIG = "big",
  MEDIUM = "medium",
  SMALL = "small",
}

interface IButton {
  onClick?: React.MouseEventHandler<HTMLButtonElement>
  disabled?: boolean
  variant?: Variants,
  size?: Size,
  inverted?: boolean,
  active?: boolean,
  icon?: IIcon,
  iconPosition?: IconPosition,
}

export const Button: React.FC<IButton> = ({
  variant,
  onClick,
  inverted,
  size,
  disabled,
  children,
  icon,
  iconPosition,
  active,
}) => {

  return (
    <style.Button
      as="button"
      onClick={onClick}
      size={size}
      inverted={inverted}
      disabled={disabled}
      variant={variant}
      active={active}
      iconPosition={iconPosition}
    >
      {iconPosition === 'left' && (icon && icon)}
      {children}
      {iconPosition === 'right' && (icon && icon)}
    </style.Button>
  );
}

Button.defaultProps = {
  variant: Variants.DEFAULT,
  iconPosition: IconPosition.LEFT,
};

interface ILinkButton extends IButton {
  link: ILinkProperties
}

export const LinkButton: React.FC<ILinkButton> = ({ link, variant, inverted, size, disabled, children, icon }) => {
  if (link.internal) {
    return (
      <style.Button
        as={Link}
        to='/'
        inverted={inverted}
        size={size}
        disabled={disabled}
        variant={variant}
      >
        {icon && icon}
        {children}
      </style.Button>
    );
  } else {
    return (
      <style.Button
        as='a'
        href={link.url}
        inverted={inverted}
        size={size}
        disabled={disabled}
        variant={variant}
      >
        {icon && icon}
        {children}
      </style.Button>
    )
  }
}

LinkButton.defaultProps = {
  variant: Variants.DEFAULT,
  iconPosition: IconPosition.LEFT,
};
