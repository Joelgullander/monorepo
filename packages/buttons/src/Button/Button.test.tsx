import * as React from 'react';

import { shallow, mount } from 'enzyme';
import { Button } from './Button';
import { ThemeProvider } from '@hiab/theme';

const ThemeProviderWrapper : React.FC = ({ children }) => (
  <ThemeProvider>
    { children }
  </ThemeProvider>
);

export const shallowWithTheme = (tree: React.ReactElement) => shallow(tree, {
  wrappingComponent: ThemeProviderWrapper
});

export const mountWithTheme = (tree: React.ReactElement) => mount(tree, {
  wrappingComponent: ThemeProviderWrapper
});


describe('<Button />', () => {
  test('renders without exploding', () => {
    const wrapper = shallowWithTheme(<Button />)
    expect(wrapper.exists()).toBe(true);
  });

  test('displays text based on props.children', () => {
    const btn = shallow(<Button>Click me</Button>);

    expect(btn.text()).toBe('Click me');

    btn.setProps({ children: 'View Product' });

    expect(btn.text()).toBe('View Product')
  });

  test('click handler', () => {
    const clickCompletionSpy = jest.fn();

    const btn = shallow(
      <Button onClick={clickCompletionSpy}>Click me</Button>
    );

    btn.simulate('click');

    expect(clickCompletionSpy).toHaveBeenCalled();
  });

  test('snapshot', () => {
    const wrapper = shallowWithTheme(<Button>Click me</Button>);

    expect(wrapper).toMatchSnapshot();
  });
});