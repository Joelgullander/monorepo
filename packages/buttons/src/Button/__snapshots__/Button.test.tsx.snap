// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> snapshot 1`] = `
<styled.div
  as="button"
  iconPosition="left"
  variant="default"
>
  Click me
</styled.div>
`;
